### PR TITLE
[extractor/telegram] Add playlist support and more metadata

### PR DESCRIPTION
**Pull request by bashonly**
- Add playlist support for Telegram when multiple videos are embedded
 - Fix IE to obey `single` query param and only extract single video from multi-video embeds
 - Add support for `--no-playlist` option
 - Improve metadata extraction (add `uploader`, `timestamp`, `duration`)
 - Eliminate unnecessary 2nd webpage request
 
Credit to @bsun0000 for bringing this issue to light and drafting an early version of this PR (#5333)

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/telegram.py [hunk 1]
```diff
@@ -1,41 +1,141 @@
...
-                r'tgme_widget_message_video_thumb"[^>]+background-image:url\(\'([^\']+)\'\)',
-                webpage_embed, 'thumbnail'),
-            'formats': formats,
+        query = {'embed': '1'}
+        if self.get_param('noplaylist'):
+            query['single'] = True
+        elif self.get_param('noplaylist') is False:
+            pass
+        elif '?single' in url:  # param 'noplaylist' is None
+            query['single'] = True
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:100]: Use utils.parse_qs and self._yes_playlist

noplaylist=None is not a thing. If `?single` is given, that should take precedence over yes/no-playlist

File: yt_dlp/extractor/telegram.py [hunk 2]
```diff
@@ -1,41 +1,141 @@
...
+                    r'<time[^>]+duration[^>]*>([\d:]+)</time>', video, 'duration', fatal=False)),
+                'thumbnail': self._search_regex(
+                    r'tgme_widget_message_video_thumb"[^>]+background-image:url\(\'([^\']+)\'\)',
+                    video, 'thumbnail', fatal=False),
+                'formats': formats,
+                **message,
+            })
+
+        if query.get('single') or len(videos) < 2:
+            return traverse_obj(videos, lambda _, x: x['id'] == msg_id, get_all=False)
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:138]: if we are getting all video info and then filtering anyway, why update `single` in the query above?

File: yt_dlp/extractor/telegram.py [hunk 3]
```diff
@@ -1,41 +1,141 @@
...
+        else:
+            return self.playlist_result(
+                videos, f'{uploader}-{msg_id}', f'{message["title"]} {msg_id}', message['description'])
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:141]: title/description should be non-fatal

File: yt_dlp/extractor/telegram.py [hunk 4]
```diff
@@ -1,41 +1,141 @@
...
+                video, 'webpage URL', fatal=False)
+            if not video_url or not webpage_url:
+                continue
+            formats = [{
+                'url': video_url,
+                'ext': 'mp4',
+            }]
+            self._sort_formats(formats)
+            videos.append({
+                'id': url_basename(webpage_url),
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:127]:
```suggestion
                'id': url_basename(webpage_url),
                'webpage_url': update_url_query(webpage_url, {'single': True}),
```

Playlist entries should have `webpage_url` if possible. When returning `url_result`, core code handles this, but when directly returning video, extractor has to manually add it

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/telegram.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/telegram.py]: `parse_qs` cannot parse a query param without a value, it returns `None`

Comment (pukkandan) [reply yt_dlp/extractor/telegram.py]: Do:
```diff
diff --git a/yt_dlp/utils.py b/yt_dlp/utils.py
index de91c66f1..0b1095f55 100644
--- a/yt_dlp/utils.py
+++ b/yt_dlp/utils.py
@@ -3067,8 +3067,8 @@ def escape_url(url):
     ).geturl()


-def parse_qs(url):
-    return urllib.parse.parse_qs(urllib.parse.urlparse(url).query)
+def parse_qs(url, **kwargs):
+    return urllib.parse.parse_qs(urllib.parse.urlparse(url).query, **kwargs)


 def read_batch_urls(batch_fd):
```
and then `parse_qs(url, keep_blank_values=True)`

File: yt_dlp/extractor/telegram.py (hunk 3 above)
Comment (bashonly) [reply yt_dlp/extractor/telegram.py:141]: `get_element_by_class` is non-fatal and the `title`/`description` keys of the `message` dict will always exist. or am I missing something?

Comment (pukkandan) [reply yt_dlp/extractor/telegram.py:141]: u r right, my bad

Comment (pukkandan) [reply yt_dlp/extractor/telegram.py:141]:
```suggestion
                videos, f'{uploader}-{msg_id}',
                format_field(message, title, f'%s {msg_id}'), message['description'])
```
 to set title=None if message[title] is None

File: yt_dlp/extractor/telegram.py (hunk 2 above)
Comment (bashonly) [reply yt_dlp/extractor/telegram.py:138]: good point. we consistently get more metadata without `single` in the embed request query, too

File: yt_dlp/extractor/telegram.py (hunk 1 above)
Comment (bashonly) [reply yt_dlp/extractor/telegram.py]: done... I *think* I've managed `?single` precedence and `_yes_playlist` output correctly

**Comment by bsun0000** [comment]
I would like to suggest to add an _extractor option_ to flip the **title** with the **description**, because title (on telegram) represents channel's name - not the title of the actual upload, while item's description is the actual title of the video. Adding multiple urls to the mpv playlist (for example) results in a plain list of the channel names.
This should not be enabled by default since this is inconvenient for a downloading - sometimes there is too much text and\\or a lot of emojis..

**Comment by bashonly** [comment]
>  _extractor option_ to flip the **title** with the **description**

I agree that it doesn't make much sense to have the channel name as the `title` for every video from a given channel. I've changed `title` to use the same message text that `description` uses.

I think an extractor-arg to toggle them would be overkill, since you can achieve the same thing with other args (`-o`, `--parse-metadata`). I also think shortening the `title` to an arbitrary length is unnecessary (and could be possibly undesirable to some users), since length can be mitigated with the output template (e.g. `%(title.0:64)s` ) or with `--trim-filenames`

File: yt_dlp/extractor/telegram.py [hunk 5]
```diff
@@ -1,41 +1,138 @@
...
     }]
 
     def _real_extract(self, url):
-        video_id = self._match_id(url)
-        webpage = self._download_webpage(url, video_id, query={'embed': 0})
-        webpage_embed = self._download_webpage(url, video_id, query={'embed': 1}, note='Downloading ermbed page')
+        channel_id, msg_id = self._match_valid_url(url).group('channel_id', 'id')
+        embed = self._download_webpage(
+            update_url_query(url, {'single': []}),  # strip 'single' from query
+            msg_id, query={'embed': '1'}, note='Downloading embed frame')
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:88]: does this not work?
```suggestion
            url msg_id, query={'embed': '1', 'single': []},
            note='Downloading embed frame')
```

File: yt_dlp/extractor/telegram.py [hunk 6]
```diff
@@ -1,41 +1,138 @@
...
-                r'tgme_widget_message_video_thumb"[^>]+background-image:url\(\'([^\']+)\'\)',
-                webpage_embed, 'thumbnail'),
-            'formats': formats,
+        def clean_text(html_class, html):
+            text = clean_html(get_element_by_class(html_class, html))
+            return text.replace('\n', ' ') if text else None
+
+        description = clean_text('tgme_widget_message_text', embed)
+        message = {
+            'title': description[:64] if description else f'Video by {channel_id}',
```
Comment (pukkandan) [inline yt_dlp/extractor/telegram.py:96]: Either use full description, or set to `''`

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Pull request merged**